packages/vue-components/src/panels/PanelBase.js

```diff
@@ -99,76 +99,85 @@ export default {
       localExpanded: false,
       localMinimized: false,
       wasRetrieverLoaded: false,
+      expandedPanelHeight: 0, // to be set dynamically
+      collapsedPanelHeight: 0, // will be >0 if show-preview mode
     };
   },
   methods: {
     toggle() {
-      if (!this.wasRetrieverLoaded && !this.localExpanded) {
-        /*
-        Let the retriever load before toggling localExpanded which triggers the animation,
-        as we need to know the scrollHeight of the content for the animation to work.
-        */
-        this.wasRetrieverLoaded = true;
-        this.$nextTick(() => {
-          this.localExpanded = true;
-        });
+      if (this.localExpanded) {
+        // COLLAPSE
+        // panel maxHeight is 'none' at this moment (as event listener set it to 'none' after expansion)
+        // thus, we need to reset the maxHeight to its current height (for collapse transition to work)
+        this.$refs.panel.style.maxHeight = `${this.$refs.panel.scrollHeight}px`;
+        // we need to reset / 'memorize' the expandedPanelHeight everytime we collapse
+        // this is because we may have expanded nested panels and the parent panel height is now different
+        this.expandedPanelHeight = this.$refs.panel.scrollHeight;
+        // need to wait for DOM update, after resetting panel maxHeight
+        setTimeout(() => {
+          // transition to collapsed panel
+          this.$refs.panel.style.maxHeight = `${this.collapsedPanelHeight}px`;
+        }, 0); // nextTick doesn't work, not sure why :O
       } else {
-        this.localExpanded = !this.localExpanded;
+        // EXPAND
+        this.$refs.panel.style.maxHeight = `${this.expandedPanelHeight}px`;
+        // we will always set maxHeight to 'none' after expansion
+        // this is to accomodate the expansion of nested panels (parent panel needs to grow)
+        // event listener will set maxHeight to none after transition ends for expansion
       }
+      this.localExpanded = !this.localExpanded;
     },
     close() {
       this.localExpanded = false;
       this.localMinimized = true;
+      // we need to reset / 'memorize' the expandedPanelHeight everytime we collapse
+      // this is because we may have expanded nested panels and the parent panel height is now different
+      this.expandedPanelHeight = this.$refs.panel.scrollHeight;
+      this.$refs.panel.style.maxHeight = `${this.collapsedPanelHeight}px`;
+      // we don't need transition for closing (change to minimized panel)
     },
     open() {
-      this.wasRetrieverLoaded = true;
-      this.$nextTick(() => {
-        this.localExpanded = true;
-      });
       this.localMinimized = false;
+      // need to wait for DOM update, after setting minimized to false
+      setTimeout(() => {
+        // transition to expanded
+        this.localExpanded = true;
+        this.$refs.panel.style.maxHeight = `${this.expandedPanelHeight}px`;
+      }, 0); // nextTick doesn't work, not sure why :O
```

Comment:
Hmm, I don't think that's actually the issue :O I discovered some weird behaviours about it, not very sure what is the actual issue. But I have documented my observations in the new commit. Perhaps you would know something about it?

Reply:
ahh ok, thanks for investigating!

this dosen't work because you are changing `this.$refs` directly (directly interfacing with the dom) instead of using vue's data interactivity (changing `localMinimized` in the previous comment / version).

> // some weird behaviours / observations here
>        // -- this.$nextTick would work if I either:
>      // 1. put a "console.log(this.$refs.panel.scrollHeight)" before it
>     // 2. OR place "if (this.$el.getBoundingClientRect().top < 0) { ... }" within the callback of nextTick
>   // (like how it is currently done below)
> // However, it would also work if i just use setTimeout instead of nextTick

Just a wild guess: these work due to referencing `scrollHeight` / `getBoundingClientRect` (e.g. `const x = this.$refs.panel.scrollHeight` would also work), which likely forces the browser to finish painting the previous `maxHeight` update first, before executing the next ``this.$refs.panel.style.maxHeight = `${this.collapsedPanelHeight}px`;``

In this case `setTimeout` is ok, though perhaps we should be using `requestAnimationFrame` instead ([3rd example here](https://css-tricks.com/using-css-transitions-auto-dimensions/)), not including the other uses of `nextTick` in (`if (!this.wasRetrieverLoaded) {` branch and in `open()`)

you may want to take a look at this comparison article between `setTimeout/Interval` and `requestAnimationFrame` as well: http://www.javascriptkit.com/javatutors/requestanimationframe.shtml